src/ft2/_layout.cpp

```diff
@@ -0,0 +1,174 @@
+  FT_Vector pos,
+  std::vector<FT_Glyph> const& glyphs,
+  std::vector<rect_t> const& rectangles)
+{
+  // Use the advance, because spaces are reported as xMin = xMax = 0, so those
+  // at the end would be ignored (OTOH, (0, 0) is always in the bbox so we
+  // don't need to special-case that).
+  auto bbox = FT_BBox{0, 0, pos.x, pos.y};
+  for (auto& glyph: glyphs) {
+    auto glyph_bbox = FT_BBox{};
+    FT_Glyph_Get_CBox(glyph, FT_GLYPH_BBOX_SUBPIXELS, &glyph_bbox);
+    bbox.xMin = std::min(bbox.xMin, glyph_bbox.xMin);
+    bbox.xMax = std::max(bbox.xMax, glyph_bbox.xMax);
+    bbox.yMin = std::min(bbox.yMin, glyph_bbox.yMin);
+    bbox.yMax = std::max(bbox.yMax, glyph_bbox.yMax);
+  }
+  for (auto& [x0, x1, y0, y1]: rectangles) {
+    bbox.xMin = std::min(bbox.xMin, FT_Pos(std::floor(x0 * 64)));
+    bbox.xMax = std::max(bbox.xMax, FT_Pos(std::ceil(x1 * 64)));
+    bbox.yMin = std::min(bbox.yMin, FT_Pos(std::floor(-y1 * 64)));
+    bbox.yMax = std::max(bbox.yMax, FT_Pos(std::ceil(-y0 * 64)));
+  }
+  return bbox;
+}
+
+Layout::Layout(
+  std::vector<FT_Glyph> const& glyphs,
+  std::vector<rect_t> const& rectangles,
+  FT_Int32 flags,
+  FT_BBox const& bbox) :
+  glyphs{glyphs}, rectangles{rectangles}, flags{flags}, bbox{bbox}
+{}
+
+Layout Layout::simple(
+  std::u32string const& string,
+  FT_Face const& face,
+  FT_Int32 flags)
+{
+  auto has_kerning = FT_HAS_KERNING(face);
+  auto previous = char32_t{};
+  auto pos = FT_Vector{};
+  auto glyphs = std::vector<FT_Glyph>{};
+  glyphs.reserve(string.size());
+  for (auto& codepoint: string) {
+    auto current = FT_Get_Char_Index(face, codepoint);
+    if (has_kerning && previous && current) {
+      auto delta = FT_Vector{};
+      FT_CHECK(
+        FT_Get_Kerning, face, previous, current, FT_KERNING_DEFAULT, &delta);
+      pos.x += delta.x;
+      pos.y += delta.y;
+    }
+    FT_CHECK(FT_Load_Glyph, face, current, flags);
+    auto glyph = FT_Glyph{};
+    FT_CHECK(FT_Get_Glyph, face->glyph, &glyph);
+    FT_CHECK(FT_Glyph_Transform, glyph, nullptr, &pos);
+    glyphs.push_back(glyph);
+    // 16.16 -> 26.6.
+    pos.x += glyph->advance.x >> 10;
+    pos.y += glyph->advance.y >> 10;
+  }
+  return {glyphs, {}, flags, compute_bbox(pos, glyphs, {})};
+}
+
+Layout Layout::manual(
+  std::vector<std::tuple<FT_Glyph, double, double>> const& positioned_glyphs,
+  std::vector<rect_t> const& rectangles,
+  FT_Int32 flags)
+{
+  auto glyphs = std::vector<FT_Glyph>{};
+  glyphs.reserve(positioned_glyphs.size());
+  for (auto& [glyph, x, y]: positioned_glyphs) {
+    auto copy = FT_Glyph{};
+    FT_CHECK(FT_Glyph_Copy, glyph, &copy);
+    auto pos =
+      FT_Vector{FT_Pos(std::round(x * 64)), FT_Pos(std::round(-y * 64))};
+    FT_Glyph_Transform(copy, nullptr, &pos);
+    glyphs.push_back(copy);
+  }
+  return {glyphs, rectangles, flags, compute_bbox({}, glyphs, rectangles)};
+}
+
+Layout::~Layout()
+{
+  if (!moved) {
+    std::for_each(glyphs.begin(), glyphs.end(), FT_Done_Glyph);
```

Comment:
If you moved out to an empty `Layout`  the `glyphs` will be empty so no need in extra handling, you can just remove the flag and leave defaulted move constructor `Layout(Layout&&) = default;`, or if you are for some reason do not need to free the glyphs (after moving to `Layout` with some glyphs) you should just vall `glyphs.clear()` in move-ctor.

Reply:
std::move will not empty the first vector (why would it? the vector is in an invalid state after the move, it may well still be the same).
And as you noticed the flag is there to avoid FT_Done_Glyph'ing twice the same glyph after a move occurred.